kep/108-Memory-bandwidth-and-LLC-aware-filtering-and-scoring/README.md

```diff
@@ -0,0 +1,145 @@
+
+## Goals
+1. Use scheduler plugin, which is the most Kubernetes native way, to implement memory bandwidth and LLC aware filtering and scoring.
+2. Leverage memory bandwidth and LLC relative metrics for filtering and scoring.
+3. Provide configurable weigths for prioritizing the metrics used in the scoring calculations.
+
+## Non-Goals
+
+## Use Cases
+
+## Terms
+LLC: last level cache.
+RDT: Intel Resource Director Technology.
+SLA: Service Level Agreement.
+MPKI: Misses Per Kilo Instructions.
+
+## Proposal
+We introduce some metrics such as memory bandwidth utilization, LLC occupation etc. At schedule stage, we can leverage these metrics to assist filting and scoring. It is hard to determine how important a resource is and how much resource is required. We did off-line analysis to different scenario and get a series of coefficient of correlation and generic usage value. The coefficient is called affinity and the generic usage value is called profile. The affinities and profiles are provided by a yaml configuration file. Detail about how the affinity and profile are used is described in the following design details section. The average value of a period of time is more meaningful than immediate value for some metrics, for instance, LLC occupancy and CPU usage. Metrics data aren't obtained from agents (such as cAdvisor) directly, instead they are stored in Prometheus first and then are retrieved on demand.
+
+## Design Details
+The filtering and scoring plugin leverages the metrics retrieved from Prometheus to determine the node candidates for running a pod. The metrics considered by the plugin at the moment are:
+
+* For filtering out oversubscribe nodes: 
+
+    * free memory available (GB)
+    * free memory bandwidth (GB/s)
+
+* For scoring:
+
+    * memory bandwidth (GB/s)
+    * memory latency (ns)
+    * Last Level Cache - l3 - Utilization (Bytes)
+    * Last Level Cache contention -- (MPKI - misses per Kilo Instructions)
+    * CPU utilization (%)
+
+the detail about filtering and scoring will be described below. 
+
+. Filtering
+
+This aspect is resposible for filtering out any nodes that would not fit our pod's profile. 
+The pod's profiles are created by analyzing top-down and finding an average scenario, which we can predict will happen on the cluster as well.
+
+In principle, in order to avoid unnecessary oversubscription, we use an *OVERPROVISIONING* factor, by default set to 2. The factor is provided by the plugin configuration file.
+
+Therefore the formulas we use are:
+
+     if (TOTAL_MEMORY_BANDWIDTH - MEMORY_BANDWIDTH_UTILIZED > OVERPROVISIONING * MEMORY_BANDWIDTH_PROFILE ):
+        return NODE_PASSED_MEMORY_BANDWITH_FILTER
+     if (FREE_MEMORY > OVERPROVISIONING * MEMORY_UTILIZATION_PROFILE ):
+        return NODE_PASSED_FREE_MEMORY_FILTER
```

Comment:
Is FREE_MEMORY a dynamic value, if so, can it potentially conflict with a Pod's resource requests?

MEMORY_UTILIZATION_PROFILE sounds like a ratio, not a concrete value, if so, isn't OVERPROVISIONING * MEMORY_UTILIZATION_PROFILE also a ratio?

Reply:
Bump. Please define these variables before using them, otherwise, readers would get confused.